### PR TITLE
Add three-phase option to OneCycleLR

### LINKED ISSUE #40362: OneCycleLR does not follow the algorithm described by the original paper
## 🐛 Bug

`torch.optim.lr_scheduler.OneCycleLR` claims to be an implementation of the schedule originally described in the paper [Super-Convergence: Very Fast Training of Neural Networks Using Large Learning Rates](https://arxiv.org/abs/1708.07120), but does not seem to match the algorithm described by the authors.

Here is a quote from that paper:

> Here we suggest a slight modification of cyclical learning rate policy for super-convergence; always use one cycle that is smaller than the total number of iterations/epochs and allow the learning rate to decrease several orders of magnitude less than the initial learning rate for the remaining iterations. We named this learning rate policy “1cycle” and in our experiments this policy allows an improvement in the accuracy.

This implies a learning rate schedule with 3 phases:

1. Increase from initial learning rate to max learning rate
2. Decrease from max learning rate to initial learning rate
3. Decrease from initial learning rate to minimum learning rate

expected

However, in its current form `OneCycleLR` only has two phases:

1. Increase from initial learning rate to max learning rate
2. Decrease from max learning rate to minimum learning rate

actual

There is a similar issue with momentum only having 2 phases instead of 3.

## To Reproduce

```python
from torch import nn
from torch import optim
from torch.optim.lr_scheduler import OneCycleLR
import matplotlib.pyplot as plt

model = nn.Linear(1, 1)
max_lr = 1.0
momentum = 0.9
epochs = 150
batches_per_epoch = 1000
optimiser = optim.SGD(model.parameters(), lr=max_lr, momentum=momentum)
scheduler = OneCycleLR(optimiser, max_lr, epochs=epochs, steps_per_epoch=batches_per_epoch,
                       pct_start=0.475, anneal_strategy='linear', cycle_momentum=True,
                       base_momentum=0.85, max_momentum=momentum, div_factor=10,
                       final_div_factor=1e4)

ys = []
for _ in range(epochs):
    for _ in range(batches_per_epoch):
        ys.append(optimiser.param_groups[0]['lr'])
        scheduler.step()
plt.title('OneCycleLR schedule')
plt.ylabel('Learning rate')
plt.xlabel('Step')
plt.plot(ys, c='red')
plt.show()
```

## Expected behavior

The plot should look like the first figure in this report ("Expected schedule"), but instead looks like the second ("OneCycleLR schedule").

## Environment

PyTorch version: 1.4.0
Is debug build: No
CUDA used to build PyTorch: 10.0

OS: Ubuntu 20.04 LTS
GCC version: (Ubuntu 9.3.0-10ubuntu2) 9.3.0
CMake version: version 3.14.0

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: GPU 0: GeForce GTX 980
Nvidia driver version: 440.64
cuDNN version: Could not collect

## Additional context

There are other 1cycle implementations/write-ups around the Internet which correctly identify that the schedule has three phases. Some examples are:

* https://sgugger.github.io/the-1cycle-policy.html
* https://github.com/titu1994/keras-one-cycle

cc @ezyang @gchanan @zou3519 @vincentqb

### PR DESCRIPTION
**Pull request by anibali**
Fixes #40362 

The new `three_phase` option provides a way of constructing schedules according to the scheme recommended in [Super-Convergence: Very Fast Training of Neural Networks Using Large Learning Rates](https://arxiv.org/abs/1708.07120).

Note that this change maintains backwards compatibility, and as a result the default behaviour of OneCycleLR remains quite counter-intuitive.

@vincentqb

### CODE REVIEW

**Comment by mjacar** [comment]
To clarify this issue a bit, as the initial author of the `OneCycleLR` implementation in pytorch, I can tell you that it is essentially a straight port of the one-cycle implementation in [fastai](https://github.com/fastai/fastai1/). For what it's worth, the whole impetus of the implementation was that I was looking to do some experiments and the way one-cycle is implemented in fastai is tightly coupled to a bunch of other fastai stuff that I didn't want or need. It was easier to just rip out the one-cycle stuff entirely.

As for why there is asymmetrical two-phase curve instead of a three-phase curve, see the [fastai reference](https://fastai1.fast.ai/callbacks.one_cycle.html): `Unpublished work has shown even better results by using only two phases: the same phase 1, followed by a second phase where we do a cosine annealing from lr_max to 0. The momentum goes from mom_min to mom_max by following the symmetric cosine (see graph a bit below).` Feel free to take that for what it's worth. If this PR is going to be merged, I think it should come with a comment that the default behavior (as of now) follows the fastai implementation (which claims better results) while the proposed alternate behavior follows the initial paper.

**Comment by anibali** [comment]
Thanks for clarifying the reasons why `OneCycleLR` is the way that it is, @mjacar. The reference to "unpublished work" seems a bit nebulous, so I think there is still value in having the option to use 1cycle as described in the original paper.

> If this PR is going to be merged, I think it should come with a comment that the default behavior (as of now) follows the fastai implementation (which claims better results) while the proposed alternate behavior follows the initial paper.

Sure, that sounds like a good idea.

**Review by vincentqb — APPROVED** [review]
Thanks for the clarification @mjacar and for the update @anibali ! LGTM, and it will avoid confusion later.
